session.go

```diff
@@ -95,52 +114,152 @@ func (s *session) Serve() error {
 		return err
 	}
 
+	s.stmts = map[string]*nodes.PrepareStmt{}
+	s.pendingStmts = map[string]*nodes.PrepareStmt{}
+	s.portals = map[string]*portal{}
 	t := protocol.NewTransport(s.Conn)
 
 	// query-cycle
 	for {
-		msg, err := t.NextFrontendMessage()
+		msg, ts, err := t.NextFrontendMessage()
 		if err != nil {
 			return err
 		}
 
-		switch v := msg.(type) {
-		case *pgproto3.Terminate:
-			s.Conn.Close()
-			return nil // client terminated intentionally
-		case *pgproto3.Query:
-			q := &query{
-				transport: t,
-				sql:       v.String,
-				queryer:   s.Server,
-				execer:    s.Server,
-			}
-			err = q.Run(s)
-			if err != nil {
-				return err
-			}
-		case *pgproto3.Describe:
-			err = t.Write(protocol.ErrorResponse(fmt.Errorf("not implemented")))
-			if err != nil {
-				return err
-			}
-		case *pgproto3.Parse:
-			err = t.Write(protocol.ErrorResponse(fmt.Errorf("not implemented")))
-			if err != nil {
-				return err
-			}
-		case *pgproto3.Bind:
-			err = t.Write(protocol.ErrorResponse(fmt.Errorf("not implemented")))
-			if err != nil {
-				return err
+		s.handleTransactionState(ts)
+		err = s.handleFrontendMessage(t, msg)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func (s *session) handleFrontendMessage(t *protocol.Transport, msg pgproto3.FrontendMessage) (err error) {
+	var res []protocol.Message
+	switch v := msg.(type) {
+	case *pgproto3.Terminate:
+		s.Conn.Close()
+		return nil // client terminated intentionally
+	case *pgproto3.Query:
+		q := &query{
+			transport: t,
+			sql:       v.String,
+			queryer:   s.Server,
+			execer:    s.Server,
+		}
+		err = q.Run(s)
+	case *pgproto3.Describe:
+		res, err = s.describe(v)
+	case *pgproto3.Parse:
+		res, err = s.prepare(v)
+	case *pgproto3.Bind:
+		res, err = s.bind(v)
+	default:
+		res = append(res, protocol.ErrorResponse(Unsupported("message type")))
+	}
+	for _, m := range res {
+		err = t.Write(m)
+		if err != nil {
+			break
+		}
+	}
+	return
+}
+
+func (s *session) handleTransactionState(state protocol.TransactionState) {
+	switch state {
+	case protocol.InTransaction, protocol.NotInTransaction:
+		// these states have no effect on session
+		break
```

Comment:
No need for a `break`, but I guess we could keep it to emphasize the emptiness of these two states😼

Reply:
that was the motivation behind this `break` :)